### PR TITLE
REF: Move pandas.core.categorical

### PR DESCRIPTION
**Pull request by TomAugspurger**
Prep for https://github.com/pandas-dev/pandas/pull/19268

### CODE REVIEW

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -0,0 +1,2331 @@
...
+        categories : index-like
+            The categories for the categorical. Items need to be unique.
+        ordered : boolean, (default False)
+            Whether or not this categorical is treated as a ordered
+            categorical. If not given, the resulting categorical will be
+            unordered.
+        """
+        try:
+            codes = np.asarray(codes, np.int64)
+        except (ValueError, TypeError):
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: Note: I changed this from `except Exception` for the linter.

File: pandas/core/categorical.py [hunk 2]
```diff
@@ -1,2331 +1,2 @@
...
-        _maxlen = 10
-        if len(self._codes) > _maxlen:
-            result = self._tidy_repr(_maxlen)
-        elif len(self._codes) > 0:
-            result = self._get_repr(length=len(self) > _maxlen)
-        else:
-            msg = self._get_repr(length=False, footer=True).replace("\n", ", ")
-            result = ('[], {repr_msg}'.format(repr_msg=msg))
-
-        return re
```
Comment (TomAugspurger) [inline pandas/core/categorical.py]: Is this unavoidable for older pickle files?

Comment (jreback) [reply pandas/core/categorical.py]: not sure what you are pointing to. but you need to have an entry in ``pandas.compat/pickle_compat.py`` to handle the pickle compat

Comment (TomAugspurger) [reply pandas/core/categorical.py]: Thanks. I thought we had something like that.

**Comment by TomAugspurger** [comment]
There sure are a lot of people importing `from pandas.core.categorical` :/ https://github.com/search?q=from+pandas.core.categorical+import&type=Code&utf8=%E2%9C%93

I think I'll add a small future warning saying to use the public API.

**Comment by jreback** [comment]
you could add a deprecating module (e.g. like tslib.py) in the to-level to make a nice transition

File: pandas/core/arrays/categorical.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/categorical.py]: +1

**Comment by jbrockmendel** [comment]
This looks pretty benign to me (I'm assuming there weren't any clandestine changes in the cut/paste).

I'd argue for calling it `vectors` instead of `arrays` so as to a) draw a clear distinction between the responsibilities of these objects vs np.array vs pd2 WIP and b) emphasize "FooVector is a vectorized version of FooScalar" (admittedly a motivation more specific to the arith/comparisons I've been focused on)

**Comment by jreback** [comment]
you prob need to change the paths in api.rst as well

**Comment by jreback** [comment]
> I'd argue for calling it vectors instead of arrays so as to a) draw a clear distinction between the responsibilities of these objects vs np.array vs pd2 WIP and b) emphasize "FooVector is a vectorized version of FooScalar" (admittedly a motivation more specific to the arith/comparisons I've been focused on)

no this is naming scheme is well-established

**Comment by TomAugspurger** [comment]
api.rst is all good. All the references via the public API.

**Comment by jreback** [comment]
not sure if you saw my comment above. I think you need a dummpy ``pandas.core.categorical.py`` which shows a FutureWarning.

**Comment by TomAugspurger** [comment]
Yeah I have one. The diff is strange. The warning is at the top, and the import is at the bottom.

**Comment by jreback** [comment]
I c, ok. can you add a test for this (maybe in test_api), and add to the deprecations list. otherwise lgtm.

**Comment by TomAugspurger** [comment]
Added a test.

**Comment by TomAugspurger** [comment]
Are we cool with merging PRs before the Mac build finishes? Travis had another outage, so they're backlogged again.

**Comment by TomAugspurger** [comment]
I think I'm going to merge PRs that are unlikely to have issues specific to the Mac build. https://travis-ci.org/pandas-dev/pandas/builds/330306709 started 2.5 hours ago, and the Mac build hasn't been scheduled yet.

I'll do cleanup if I *do* break anything :)

**Pull request merged**